src/consensus/aft/raft.h

```diff
@@ -1380,57 +1519,254 @@ namespace aft
+                  scheduler_thread, std::move(msg));
+              },
+              this,
+              std::move(ds),
+              state->last_idx,
+              state->commit_idx,
+              threading::get_current_thread_id());
+
+            async_executor.increment_pending();
+            threading::ThreadMessaging::thread_messaging.add_task(
+              threading::ThreadMessaging::get_execution_thread(
+                ++next_exec_thread),
+              std::move(tmsg));
           }
+          else
+          {
+            executor->commit_replayed_request(
+              ds->get_request(),
+              request_tracker,
+              state->last_idx,
+              state->commit_idx,
+              ds->get_max_conflict_version());
+          }
+          break;
+        }
+
+        case kv::ApplyResult::PASS_SNAPSHOT_EVIDENCE:
+        {
+          break;
+        }
+
+        default:
+        {
+          throw std::logic_error("Unknown ApplyResult value");
+        }
+      }
+      return AsyncExecutionResult::COMPLETE;
+    }
+
+    AsyncExecutionResult execute_append_entries_async(
+      std::unique_ptr<threading::Tmsg<AsyncExecution>>& msg)
+    {
+      // This function is responsible for selection the next batch of
+      // transactions we can execute concurrently and then starting said
+      // execution. If there are more pending entries after we select the batch
+      // we will return to this function.
+      std::vector<
+        std::tuple<std::unique_ptr<kv::AbstractExecutionWrapper>, kv::Version>>&
+        append_entries = msg->data.append_entries;
+      AppendEntries& r = msg->data.r;
+      bool confirm_evidence = msg->data.confirm_evidence;
+      async_exec_msg = std::move(msg);
+      async_executor.execute_as_far_as_possible(state->last_idx);
+      while (async_exec_msg->data.next_append_entry_index !=
+             append_entries.size())
+      {
+        auto& [ds, i] =
+          append_entries[async_exec_msg->data.next_append_entry_index];
+        if (!async_executor.should_exec_next_append_entry(
+              ds->support_async_execution(), ds->get_max_conflict_version()))
+        {
+          return AsyncExecutionResult::PENDING;
+        }
+
+        ++async_exec_msg->data.next_append_entry_index;
+        state->last_idx = i;
+
+        kv::ApplyResult apply_result = ds->apply();
+        if (
+          process_async_execution(apply_result, ds, i, r) ==
+          AsyncExecutionResult::PENDING)
+        {
+          return AsyncExecutionResult::PENDING;
         }
       }
 
+      if (async_executor.execution_status() == AsyncExecutionResult::COMPLETE)
+      {
+        execute_append_entries_finish(confirm_evidence, r);
+        return AsyncExecutionResult::COMPLETE;
+      }
+      return AsyncExecutionResult::PENDING;
+    }
+
+    bool execute_append_entries_finish(bool confirm_evidence, AppendEntries& r)
```

Comment:
Not clear here where `confirm_evidence` comes from, and reading the code it suggests we will `commit_if_possible` _without_ confirming, if this bool can be set to false. I think it could be removed, and we can always call `view_change_tracker->check_evidence(r.term)`? That would be a nop (I've already checked and I'm happy) for subsequent calls in the same term, but the first call (per-term) it does the real evidence check.